src/Build/BackEnd/Components/Logging/LoggingService.cs

```diff
@@ -515,6 +515,40 @@ public bool HasBuildSubmissionLoggedErrors(int submissionId)
             return _buildSubmissionIdsThatHaveLoggedErrors?.Contains(submissionId) == true;
         }
 
+        /// <summary>
+        /// Returns a hashset of warnings to be logged as errors for the specified project instance ID.
+        /// Note that WarningsAsMessages takes priority over WarningsAsErrors and are excluded from the set.
+        ///
+        /// If all warnings to be treated as errors should also be treated as messages, return null otherwise
+        /// all warnings will be treated as errors. <see cref="RequestBuilder.ConfigureWarningsAsErrorsAndMessages()"/>
+        /// </summary>
+        /// <param name="context">The build context through which warnings will be logged as errors.</param>
+        /// <returns>A Hashset containing warning codes that should be treated as errors.</returns>
+        public HashSet<string> GetWarningsToBeLoggedAsErrorsByProject(BuildEventContext context)
+        {
+            if(_warningsAsErrorsByProject == null)
+            {
+                return null;
+            }
+
+            int key = GetWarningsAsErrorOrMessageKey(context);
+
+            HashSet<string> warningsAsErrorsExcludingMessages = new HashSet<string>(_warningsAsErrorsByProject?[key]);
+
+            if(_warningsAsMessagesByProject != null)
+            {
+                warningsAsErrorsExcludingMessages.ExceptWith(_warningsAsMessagesByProject[key]);
+
+                // A non-null empty set means all warnings are errors. Avoid this.
```

Comment:
This contract seems unclear. As a rule, null or empty collections shouldn't have different semantic meanings, otherwise it's very confusing. If you need to, perhaps make this explicit and either add a bool out parameter or return a tuple or a data type?

Reply:
I generally agree, but this was written with the intent of matching the current implementation (which I'm open to changing).

See these files:
1. [Line 1582 of LoggingService.cs](https://github.com/dotnet/msbuild/blob/master/src/Build/BackEnd/Components/Logging/LoggingService.cs#L1586)
2. [Line 1242 of RequestBuilder.cs](https://github.com/dotnet/msbuild/blob/master/src/Build/BackEnd/Components/RequestBuilder/RequestBuilder.cs#L1242)

We talked a bit about it in PR review and didn't come to a strong conclusion for any particular side.

It's worth mentioning that if we switch IBE8 from having a hashset to having a method that returns true or false, we can change the underlying implementation with less hassle. Whereas if we merge this and decide to add a type or extra information to determine warningsaserrors, we'd need to add an IBE9 or modify IBE8.

/cc: @jeffkl as the original author